### ISSUE TITLE
paddleseg人像分割模型问题

### ISSUE THREAD

**Issue by wsy1991** [PaddlePaddle/PaddleSeg#916]
logit = self.output_layer(decode_conv, self.num_classes)
        #print ('logit', logit)
        if self.num_classes == 1:
            out = sigmoid_to_softmax(logit)
            out = fluid.layers.transpose(out, [0, 2, 3, 1])
        else:
            out = fluid.layers.transpose(logit, [0, 2, 3, 1])
        
        pred = fluid.layers.argmax(out, axis=3)
        pred = fluid.layers.unsqueeze(pred, axes=[1])  
        
        if self.mode == 'train':
            label = inputs['label']
            mask = label != self.ignore_index
            return self._get_loss(logit, label, mask)

        else:
            if self.num_classes == 1:
                logit = sigmoid_to_softmax(logit)
            else:
                logit = fluid.layers.softmax(logit, axis=1)
            return pred, logit

        return logit

在paddleseg中人像分割模型中(shufflenet，hrnet，deeplab)中都有这一段代码。我check了下，pred用于评估，softmax出来的logit用于预测部署。但这样有一个问题，最终用于部署的模型，会同时包含argmax和softmax操作，我profiler了shufflenet_slim的模型的结果是，这两个操作都非常耗时，影响模型推理速度。profiler结果如下：

-------------------------       Event Summary       -------------------------

Event                            Calls       Total       Min.        Max.        Ave.        Ratio.      
thread0::softmax                 1           5.58693     5.58693     5.58693     5.58693     0.154435    
thread0::arg_max                 1           3.61234     3.61234     3.61234     3.61234     0.0998527   
thread0::bilinear_interp         1           2.16216     2.16216     2.16216     2.16216     0.0597665   
thread0::conv2d_transpose        1           1.23317     1.23317     1.23317     1.23317     0.0340873   
thread0::pool2d                  1           0.846578    0.846578    0.846578    0.846578    0.0234012   
thread0::elementwise_add         1           0.838149    0.838149    0.838149    0.838149    0.0231682   
thread0::feed                    1           0.773218    0.773218    0.773218    0.773218    0.0213734   
thread0::depthwise_conv2d        16          5.19427     0.113885    2.89118     0.324642    0.143581    
thread0::conv2d                  30          5.72721     0.031944    2.77596     0.190907    0.158312    
thread0::transpose2              13          1.82909     0.013437    0.975365    0.140699    0.0505598   
thread0::concat                  13          1.54642     0.010266    0.956869    0.118955    0.0427463   
thread0::batch_norm              46          3.1624      0.020871    0.976829    0.0687479   0.0874154   
thread0::split                   10          0.426008    0.011176    0.296092    0.0426008   0.0117758   
thread0::slice                   37          1.48232     0.008472    1.04049     0.0400628   0.0409745   
thread0::unsqueeze2              1           0.039595    0.039595    0.039595    0.039595    0.00109449  
thread0::reshape2                24          0.718796    0.012341    0.336114    0.0299498   0.019869    
thread0::relu                    30          0.536617    0.006532    0.151702    0.0178872   0.0148332   
thread0::fill_constant           36          0.371151    0.007417    0.020437    0.0103097   0.0102594   
thread0::shape                   13          0.090283    0.005592    0.013242    0.00694485  0.00249561

**Comment by LutaoChu** [comment]
预测部署难道不需要通过argmax获取类别吗？
如果确实不需要的话，可以把pred相关的代码删掉导出新的模型

**Comment by wsy1991** [comment]
> 预测部署难道不需要通过argmax获取类别吗？
> 如果确实不需要的话，可以把pred相关的代码删掉导出新的模型

这样的话代码会报错。
我的意思是，是否有可能去掉softmax，因为真的很影响推理速度

**Comment by LutaoChu** [comment]
噢噢，肯定是可以去掉softmax的。
这样的话你可以用pred，pred不需要过softmax。

**Comment by LutaoChu** [comment]
![image](https://user-images.githubusercontent.com/30695251/112429395-8e1fb780-8d77-11eb-800b-6a20e512d719.png)
具体来说，注释掉 /legacy/contrib/HumanSeg/models/humanseg.py里的这一行试试

**Comment by wsy1991** [comment]
但是deploy下infer.py中用的应该是logit吧
 output_data = self.predictor.run(input_data)[1]
            output_data = output_data.as_ndarray()
Humanseg中bp_replace.py中用的应该也是logit啊
bp_replace.py中结合光流，看上去需要用概率。
我用netron查看了下模型结构，softmax是包含在模型中的

**Comment by LutaoChu** [comment]
deploy下infer.py用的是pred，
```
output_data = self.predictor.run(input_data)[0]
```
你应该自己修改了代码吧

bg_replace.py用的是logit，需要用概率

**Comment by wuyefeilin** [comment]
> 但是deploy下infer.py中用的应该是logit吧
> output_data = self.predictor.run(input_data)[1]
> output_data = output_data.as_ndarray()
> Humanseg中bp_replace.py中用的应该也是logit啊
> bp_replace.py中结合光流，看上去需要用概率。
> 我用netron查看了下模型结构，softmax是包含在模型中的

bg_replace.py里面用的是概率哈https://github.com/PaddlePaddle/PaddleSeg/blob/release/v2.0/legacy/contrib/HumanSeg/bg_replace.py#L83

**Comment by wsy1991** [comment]
哦对。bp_replace用的是概率(softmax)。因为后面需要用光流。但是光流和softmax都特别耗时，其中，softmax占了模型推理40%的时间。我在想，有啥方法能够不用softmax或者减少光流推理时间呢

**Comment by wuyefeilin** [comment]
> 哦对。bp_replace用的是概率(softmax)。因为后面需要用光流。但是光流和softmax都特别耗时，其中，softmax占了模型推理40%的时间。我在想，有啥方法能够不用softmax或者减少光流推理时间呢

你可以尝试把softmax 去除，重新导出模型，然后在部署的时候自己加上softmax，看看这么操作是否能节省时间

**Comment by wsy1991** [comment]
区别在哪嘛，哥哥。想哭.jpg

**Comment by wsy1991** [comment]
请问下，如果有softmax,有啥方法可以加快softmax推理嘛@wuyefeilin

**Comment by wsy1991** [comment]
> > 哦对。bp_replace用的是概率(softmax)。因为后面需要用光流。但是光流和softmax都特别耗时，其中，softmax占了模型推理40%的时间。我在想，有啥方法能够不用softmax或者减少光流推理时间呢
> 
> 你可以尝试把softmax 去除，重新导出模型，然后在部署的时候自己加上softmax，看看这么操作是否能节省时间

尝试过了，并不会节省时间

**Comment by Huihuihh** [comment]
我想咨询下，PaddleSeg推理，如何获取缺陷类别

**Comment by LutaoChu** [comment]
缺陷类别是？
